src/ansi.rs

```diff
@@ -0,0 +1,1849 @@
+        let end = &self.state.sync_state.buffer[offset..];
+
+        let sync_buffer_size: usize;
+        sync_buffer_size = SYNC_BUFFER_SIZE;
+
+        // Check for extension/termination of the synchronized update.
+        if end == SYNC_START_ESCAPE_START {
+            self.state.sync_state.pending_dcs = Some(Dcs::SyncStart);
+        } else if end == SYNC_END_ESCAPE_START || len >= sync_buffer_size - 1 {
+            self.state.sync_state.pending_dcs = Some(Dcs::SyncEnd);
+        }
+    }
+
+    /// Parse the DCS termination sequence for synchronized updates.
+    #[cfg(feature = "alloc")]
+    fn advance_sync_dcs_end<H, W>(&mut self, handler: &mut H, writer: &mut W, byte: u8)
+    where
+        H: Handler<W>,
+    {
+        match byte {
+            // Ignore DCS passthrough characters.
+            0x00..=0x17 | 0x19 | 0x1c..=0x7f | 0xa0..=0xff => (),
+            // Cancel the DCS sequence.
+            0x18 | 0x1a | 0x80..=0x9f => self.state.sync_state.pending_dcs = None,
+            // Dispatch on ESC.
+            0x1b => match self.state.sync_state.pending_dcs.take() {
+                Some(Dcs::SyncStart) => {
+                    self.state.sync_state.timeout = Some(
+                        self.state.sync_state.time_provider.now() + SYNC_UPDATE_TIMEOUT.as_millis(),
+                    );
+                    self.state.sync_state.in_sync = true;
+                },
+                Some(Dcs::SyncEnd) => self.stop_sync(handler, writer),
+                None => (),
+            },
+        }
+    }
+}
+
+/// Helper type that implements `Perform`.
+///
+/// Processor creates a Performer when running advance and passes the Performer
+/// to `Parser`.
+struct Performer<'a, H: Handler<W>, W, T: TimeProvider> {
+    state: &'a mut ProcessorState<T>,
+    handler: &'a mut H,
+    writer: &'a mut W,
+}
+
+impl<'a, H: Handler<W> + 'a, W, T: TimeProvider> Performer<'a, H, W, T> {
+    /// Create a performer.
+    #[inline]
+    pub fn new<'b>(
+        state: &'b mut ProcessorState<T>,
+        handler: &'b mut H,
+        writer: &'b mut W,
+    ) -> Performer<'b, H, W, T> {
+        Performer { state, handler, writer }
+    }
+}
+
+/// Type that handles actions from the parser.
+///
+/// XXX Should probably not provide default impls for everything, but it makes
+/// writing specific handler impls for tests far easier.
+pub trait Handler<W> {
+    /// OSC to set window title.
+    #[allow(unused_variables)]
+    fn set_title(&mut self, params: Option<&[&[u8]]>) {
+        #[cfg(feature = "alloc")]
+        {
+            use alloc::borrow::ToOwned;
+
+            self.set_title_utf(params.map(|params| {
+                params[1..]
+                    .iter()
+                    .flat_map(|x| str::from_utf8(x))
+                    .collect::<Vec<&str>>()
+                    .join(";")
+                    .trim()
+                    .to_owned()
+            }))
+        }
```

Comment:
What exactly is the point of this? Having separate `set_title` functions just makes things complicated and easy to screw up. Performing some string manipulation of params in the `Handler` is completely inappropriate.